lib/rspecq/worker.rb

```diff
@@ -163,39 +163,33 @@ def try_publish_queue!(queue)
       puts "Published queue (size=#{queue.publish(jobs)})"
     end
 
+    # Split a list of spec files into their individual examples.
+    #
     # NOTE: RSpec has to load the files before we can split them as individual
     # examples. In case a file to be splitted fails to be loaded
     # (e.g. contains a syntax error), we return the files unchanged, thereby
     # falling back to scheduling them as whole files. Their errors will be
     # reported in the normal flow when they're eventually picked up by a worker.
     def files_to_example_ids(files)
-      cmd = "DISABLE_SPRING=1 bundle exec rspec --dry-run --format json #{files.join(' ')}"
-      out, err, cmd_result = Open3.capture3(cmd)
-
-      if !cmd_result.success?
-        rspec_output = begin
-                         JSON.parse(out)
-                       rescue JSON::ParserError
-                         out
-                       end
-
-        log_event(
-          "Failed to split slow files, falling back to regular scheduling.\n #{err}",
-          "error",
-          rspec_stdout: rspec_output,
-          rspec_stderr: err,
-          cmd_result: cmd_result.inspect,
-        )
+      out = StringIO.new
 
-        pp rspec_output
+      reset_rspec_state!
 
-        return files
-      end
+      dry_run_before = RSpec.configuration.dry_run
 
-      JSON.parse(out)["examples"].map { |e| e["id"] }
+      RSpec.configuration.add_formatter(RSpec::Core::Formatters::JsonFormatter.new(out))
+      RSpec.configuration.files_or_directories_to_run = files_or_dirs_to_run
+      RSpec.configuration.dry_run = true
+
+      opts = RSpec::Core::ConfigurationOptions.new(files)
+      result = RSpec::Core::Runner.new(opts).run($stderr, $stdout)
+      return files if result != 0
+      JSON.parse(out.string)["examples"].map { |e| e["id"] }
+    ensure
+      RSpec.configuration.dry_run = dry_run_before
+      reset_rspec_state!
```

Comment:
Explain, why do you need to set the original value for the dry_run (only)?

Reply:
Because that's the one we're messing with (we set it to `true` so that we can fetch the example ids without executing any tests).